src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs

```diff
@@ -698,20 +972,44 @@
                 AreConstraintTypesSubset(substitutedTypes2, substitutedTypes1, typeParameter1);
         }
 
-        public static bool HaveSameNullabilityInConstraints(TypeParameterSymbol typeParameter1, TypeMap typeMap1, TypeParameterSymbol typeParameter2, TypeMap typeMap2)
+        public static bool HaveSameNullabilityInConstraints(TypeParameterSymbol typeParameter1, TypeMap? typeMap1, TypeParameterSymbol typeParameter2, TypeMap? typeMap2, TypeCompareKind typeComparison)
         {
-            if (!typeParameter1.IsValueType)
+            if (!typeParameter1.IsValueType
+                && !hasSameLevelNullability(typeParameter1, typeParameter2, typeComparison))
+            {
+                return false;
+            }
+
+            return HaveSameTypeConstraints(typeParameter1, typeMap1, typeParameter2, typeMap2, SymbolEqualityComparer.Create(typeComparison));
+
+            static bool hasSameLevelNullability(TypeParameterSymbol typeParameter1, TypeParameterSymbol typeParameter2, TypeCompareKind typeComparison)
             {
                 bool? isNotNullable1 = typeParameter1.IsNotNullable;
                 bool? isNotNullable2 = typeParameter2.IsNotNullable;
+
+                if ((typeComparison & TypeCompareKind.ObliviousNullableModifierMatchesAny) != 0)
+                {
+                    if (!isNotNullable1.HasValue || !isNotNullable2.HasValue)
```

Comment:
>if (!isNotNullable1.HasValue || !isNotNullable2.HasValue) [](http://example.com/codeflow?start=20&length=57)

Consider pulling this common condition out of both branches of the ```if```. I think the code will be easier to follow and it will be obvious that ```isNotNullable1.HasValue && isNotNullable2.HasValue``` is always true in the ```if``` below.